File: 2001-2500/2392_buildMatrix.py
class Solution:
    def buildMatrix(self, k: int, rowConditions: List[List[int]], colConditions: List[List[int]]) -> List[List[int]]:
        # Classic topological sort problem. Both for rows and columns, Kahn's Algorithm will be applied.
        # Topologically sorted row and columns will show the correct positions of numbers (1 to k) in matrix (k x k)

        rows = self.topologicalSort(k, rowConditions)
        cols = self.topologicalSort(k, colConditions)

        if rows == [] or cols == []:
            return []

        indices = {i: [] for i in range(1, k + 1)}

        for idx, row in enumerate(rows):
            indices[row].append(idx)

        for idx, col in enumerate(cols):
            indices[col].append(idx)

        matrix = [[0] * k for _ in range(k)]

        for num, [row, col] in indices.items():
            matrix[row][col] = num

        return matrix

    def topologicalSort(self, n, edges):
        graph = {i: [] for i in range(1, n + 1)}
        in_degree = {i: 0 for i in range(1, n + 1)}

        for start, end in edges:
            graph[start].append(end)
            in_degree[end] += 1

        queue = deque([node for node, degree in in_degree.items() if not degree])

        toposort = []

        while queue:
            node = queue.popleft()
            toposort.append(node)

            for neighbor in graph[node]:
                in_degree[neighbor] -= 1

                if in_degree[neighbor] == 0:
                    queue.append(neighbor)

        return toposort if len(toposort) == n else []

File: 2001-2500/2392_buildMatrix.java
class Solution {
    public int[][] buildMatrix(int k, int[][] rowConditions, int[][] colConditions) {
        int[][] coordinates = new int[k][2];
        int[][] matrix = new int[k][k];
        List<Integer> toposortedRow = toposort(k, rowConditions);
        List<Integer> toposortedCol = toposort(k, colConditions);

        if (toposortedRow.size() != k || toposortedCol.size() != k) {
            return new int[0][0];
        }

        for (int i = 0; i < k; i++) {
            int rowVal = toposortedRow.get(i);
            int colVal = toposortedCol.get(i);

            coordinates[rowVal][0] = i;
            coordinates[colVal][1] = i;
        }

        for (int i = 0; i < k; i++) {
            int[] coordinate = coordinates[i];
            int r = coordinate[0];
            int c = coordinate[1];

            matrix[r][c] = i + 1;
        }

        return matrix;
    }

    public List<Integer> toposort(int k, int[][] conditions) {
        ArrayList<Integer>[] graph = new ArrayList[k];
        int[] indegree = new int[k];
        Queue<Integer> queue = new LinkedList<>();
        List<Integer> result = new ArrayList<>();

        for (int i = 0; i < k; i++) {
            graph[i] = new ArrayList<>();
        }

        for (int[] condition: conditions) {
            int u = condition[0] - 1, v = condition[1] - 1;
            graph[u].add(v);
            indegree[v]++;
        }

        for (int i = 0; i < k; i++) {
            if (indegree[i] == 0) {
                queue.add(i);
            }
        }

        while (!queue.isEmpty()) {
            int node = queue.remove();
            result.add(node);

            for (int child: graph[node]) {
                indegree[child]--;

                if (indegree[child] == 0) {
                    queue.add(child);
                }
            }
        }

        return result;
    }
}
